test/unit/src/utils_test.dart

```diff
@@ -128,4 +128,136 @@ void main() {
       expect(allMatches[1].group(1), 'bar');
     });
   });
+
+  group('doesVersionPinDependency', () {
+    test('is false for any', () {
+      expect(doesVersionPinDependency('any'), isFalse);
+      expect(doesVersionPinDependency("'any'"), isFalse);
+      expect(doesVersionPinDependency('"any"'), isFalse);
+    });
+
+    test('is true for empty', () {
+      expect(doesVersionPinDependency('>0.0.0 <0.0.0'), isTrue);
+      expect(doesVersionPinDependency('">0.0.0 <0.0.0"'), isTrue);
+      expect(doesVersionPinDependency("'>0.0.0 <0.0.0'"), isTrue);
+    });
+
+    test('with carat notation is false', () {
+      expect(doesVersionPinDependency('^0.0.1'), isFalse);
+      expect(doesVersionPinDependency('"^0.0.19"'), isFalse);
+      expect(doesVersionPinDependency("'^0.0.36'"), isFalse);
+
+      expect(doesVersionPinDependency('^0.2.4'), isFalse);
+      expect(doesVersionPinDependency('"^0.23.456"'), isFalse);
+      expect(doesVersionPinDependency("'^0.23.456'"), isFalse);
+
+      expect(doesVersionPinDependency('^1.2.4'), isFalse);
+      expect(doesVersionPinDependency('"^1.23.456"'), isFalse);
+      expect(doesVersionPinDependency("'^1.23.456'"), isFalse);
+    });
+
+    test('is true for 1.2.3', () {
+      expect(doesVersionPinDependency('1.23.456'), isTrue);
+      expect(doesVersionPinDependency('"1.23.456"'), isTrue);
+      expect(doesVersionPinDependency("'1.23.456'"), isTrue);
+    });
+
+    test('is true for explicit upper bound <=', () {
+      expect(doesVersionPinDependency('">=1.2.3 <=4.0.0"'), isTrue);
+      expect(doesVersionPinDependency("'>=1.2.3 <=4.0.0'"), isTrue);
+      expect(doesVersionPinDependency('"<=4.0.0"'), isTrue);
+      expect(doesVersionPinDependency("'<=4.0.0'"), isTrue);
+    });
+
+    group('is true if upper bound blocks patch or minor updates', () {
+      test('when version starts with 0', () {
+        expect(doesVersionPinDependency('">=0.2.3 <0.5.6"'), isTrue);
+        expect(doesVersionPinDependency("'>=0.2.3 <0.5.6'"), isTrue);
+        expect(doesVersionPinDependency('"<0.5.6"'), isTrue);
+        expect(doesVersionPinDependency("'<0.5.6'"), isTrue);
+      });
+
+      test('when version starts with nonzero', () {
+        // blocks minor updates
+        expect(doesVersionPinDependency('">=1.2.3 <4.5.0"'), isTrue);
+        expect(doesVersionPinDependency("'>=1.2.3 <4.5.0'"), isTrue);
```

Comment:
For future-proofing changes to this logic, could we add test cases for blocking minor/patch updates for the same major/minor versions?
```dart
expect(doesVersionPinDependency("'>=1.2.3 <1.2.4'"), isTrue);
expect(doesVersionPinDependency("'>=1.3.0 <1.4.0'"), isTrue);
```